File: voiceping-sdk/src/main/java/com/smartwalkie/voicepingsdk/VoicePing.java
package com.smartwalkie.voicepingsdk;

import android.content.Context;
import android.media.AudioFormat;
import android.provider.Settings;

import com.smartwalkie.voicepingsdk.callbacks.ConnectCallback;
import com.smartwalkie.voicepingsdk.callbacks.DisconnectCallback;
import com.smartwalkie.voicepingsdk.listeners.ChannelListener;
import com.smartwalkie.voicepingsdk.listeners.OutgoingTalkCallback;
import com.smartwalkie.voicepingsdk.models.AudioParam;
import com.smartwalkie.voicepingsdk.models.local.VoicePingPrefs;

import java.util.HashMap;
import java.util.Map;

/**
 * Main class of VoicePing.
 */
public class VoicePing {

    private Context mContext;
    private Player mPlayer;
    private Connection mConnection;
    private Recorder mRecorder;

    private VoicePing(Context context, String serverUrl, AudioParam audioParam) {
        mContext = context;
        mPlayer = new Player(context, audioParam);
        mPlayer.start();
        mConnection = new Connection(context, serverUrl, mPlayer);
        mRecorder = new Recorder(context, mConnection, audioParam);
        mConnection.setOutgoingAudioListener(mRecorder);
        VoicePingPrefs.getInstance(context).putServerUrl(serverUrl);
    }

    /**
     * Instantiate VoicePing. VoicePing instance is the most important instance in this SDK.
     * The developer that use this SDK can do other processes using VoicePing instance.
     *
     * @param context Application Context
     * @param serverUrl The URL of VoicePing SDK server
     * @return VoicePing instance with default parameters
     */
    public static VoicePing init(Context context, String serverUrl) {
        return new Builder().buildAndInit(context, serverUrl);
    }

    /**
     * Instantiate VoicePing.Builder class.
     *
     * @return VoicePing.Builder instance that can be used to build VoicePing instance with custom
     * parameters
     */
    public static Builder newBuilder() {
        return new Builder();
    }

    /**
     * Connect to server. This method can be assumed as sign in to server. After the user connected
     * to server, the user can then receive PTT from any other user using private channel.
     *
     * @param userId User ID or Username
     * @param callback Callback
     */
    public void connect(String userId, ConnectCallback callback) {
        Map<String, String> props = new HashMap<>();
        props.put("user_id", userId);
        VoicePingPrefs.getInstance(mContext).putUserId(userId);
        props.put("DeviceId", Settings.Secure
                .getString(mContext.getContentResolver(), Settings.Secure.ANDROID_ID));
        mConnection.connect(props, callback);
    }

    /**
     * Disconnect from server. This method can be assumed as sign out from server. After
     * disconnected from server, the user will not receive any incoming message.
     *
     * @param callback Callback
     */
    public void disconnect(DisconnectCallback callback) {
        mConnection.disconnect(callback);
    }

    /**
     * Set ChannelListener to VoicePing to do some advanced techniques to the audio data.
     *
     * @param channelListener Listener
     */
    public void setChannelListener(ChannelListener channelListener) {
        mPlayer.setChannelListener(channelListener);
    }

    /*public void subscribe(String channelId, int channelType) {

    }

    public void unsubscribe(String channelId, int channelType) {

    }*/

    /**
     * Start talking using VoicePing's PTT functionality.
     *
     * @param receiverId Receiver ID for private channel, or Group ID for group channel
     * @param channelType ChannelType.PRIVATE or ChannelType.GROUP
     */
    public void startTalking(String receiverId, int channelType, OutgoingTalkCallback callback) {
        mRecorder.startTalking(receiverId, channelType, callback);
    }

    /**
     * Stop talking.
     */
    public void stopTalking() {
        mRecorder.stopTalking();
    }

    public void subscribe(String groupId) {
        String userId = VoicePingPrefs.getInstance(mContext).getUserId();
        mConnection.send(MessageHelper.createSubscribeMessage(userId, groupId));
    }

    public void unsubscribe(String groupId) {
        String userId = VoicePingPrefs.getInstance(mContext).getUserId();
        mConnection.send(MessageHelper.createUnsubscribeMessage(userId, groupId));
    }

    /**
     * Builder class to instantiate VoicePing with custom parameters
     */
    public static final class Builder {

        private boolean isUsingOpusCodec;
        private int sampleRate;
        private int frameSize;
        private int channelSize;
        private int bufferSizeFactor;
        private int channelInConfig;
        private int channelOutConfig;
        private int audioFormat;

        private Builder() {
            isUsingOpusCodec = true;
            sampleRate = 16000;
            frameSize = 960;
            channelSize = 1;
            bufferSizeFactor = 2;
            channelInConfig = AudioFormat.CHANNEL_IN_MONO;
            channelOutConfig = AudioFormat.CHANNEL_OUT_MONO;
            audioFormat = AudioFormat.ENCODING_PCM_16BIT;
        }

        /**
         * Set whether the SDK will use Opus as audio codec or not. The default value is true.
         *
         * @param usingOpusCodec true for using Opus Codec, false otherwise
         * @return Builder
         */
        public Builder setUsingOpusCodec(boolean usingOpusCodec) {
            isUsingOpusCodec = usingOpusCodec;
            return this;
        }

        /**
         * Set sample rate of the audio data. The default value is 16000.
         *
         * @param sampleRate Sample rate in Hz. Max 48000 Hz
         * @return Builder
         */
        public Builder setSampleRate(int sampleRate) {
            this.sampleRate = sampleRate;
            return this;
        }

        /**
         * Set frame size of the audio data. The default value is 960.
         *
         * @param frameSize Frame size
         * @return Builder
         */
        public Builder setFrameSize(int frameSize) {
            this.frameSize = frameSize;
            return this;
        }

        /**
         * Set channel size of the audio data. The default value is 1.
         *
         * @param channelSize Channel size. 1 for Mono, 2 for Stereo
         * @return Builder
         */
        public Builder setChannelSize(int channelSize) {
            this.channelSize = channelSize;
            return this;
        }

        /**
         * Set buffer size factor of the audio data. The default value is 2.
         *
         * @param bufferSizeFactor Buffer size in bytes
         * @return Builder
         */
        public Builder setBufferSizeFactor(int bufferSizeFactor) {
            this.bufferSizeFactor = bufferSizeFactor;
            return this;
        }

        /**
         * Set Channel-In configuration of the audio data. The default value is
         * AudioFormat.CHANNEL_IN_MONO
         *
         * @param channelInConfig Channel-In config
         * @return Builder
         */
        public Builder setChannelInConfig(int channelInConfig) {
            this.channelInConfig = channelInConfig;
            return this;
        }

        /**
         * Set Channel-Out configuration of the audio data. The default value is
         * AudioFormat.CHANNEL_OUT_MONO
         *
         * @param channelOutConfig Channel-Out config
         * @return Builder
         */
        public Builder setChannelOutConfig(int channelOutConfig) {
            this.channelOutConfig = channelOutConfig;
            return this;
        }

        /**
         * Set audio format of the audio data. The default value is AudioFormat.ENCODING_PCM_16BIT
         *
         * @param audioFormat Audio format
         * @return Builder
         */
        public Builder setAudioFormat(int audioFormat) {
            this.audioFormat = audioFormat;
            return this;
        }

        /**
         * Build the Builder instance to be VoicePing instance, and initiate it.
         *
         * @param context Application context
         * @param serverUrl The URL of VoicePing SDK server
         * @return VoicePing instance
         */
        public VoicePing buildAndInit(Context context, String serverUrl) {
            AudioParam audioParam = new AudioParam(isUsingOpusCodec, sampleRate, frameSize,
                    channelSize, bufferSizeFactor, channelInConfig, channelOutConfig, audioFormat);
            return new VoicePing(context, serverUrl, audioParam);
        }
    }
}

File: voiceping-sdk/src/main/java/com/smartwalkie/voicepingsdk/VoicePing.kt
package com.smartwalkie.voicepingsdk

import android.content.Context
import android.os.Build
import android.os.HandlerThread
import com.smartwalkie.voicepingsdk.callback.ConnectCallback
import com.smartwalkie.voicepingsdk.callback.DisconnectCallback
import com.smartwalkie.voicepingsdk.listener.ConnectionStateListener
import com.smartwalkie.voicepingsdk.listener.IncomingTalkListener
import com.smartwalkie.voicepingsdk.listener.OutgoingTalkCallback
import com.smartwalkie.voicepingsdk.model.AudioParam

/**
 * Main class of VoicePing.
 */
object VoicePing {
    private const val DEFAULT_SERVER_URL: String = "wss://router-lite.voiceping.info"
    private lateinit var audioParam: AudioParam
    private lateinit var player: Player
    private lateinit var connection: Connection
    private lateinit var recorder: Recorder

    private var userId: String? = null
    private var company: String? = null

    /**
     * Initialize VoicePing. This process should be done in onCreate of your Application class
     *
     * @param context   Application Context
     */
    fun init(context: Context) {
        init(context, AudioParam.Builder().build())
    }

    /**
     * Initialize VoicePing. This process should be done in onCreate of your Application class
     *
     * @param context   Application Context
     * @param audioParam Audio parameters
     */
    fun init(context: Context, audioParam: AudioParam) {
        this.audioParam = audioParam
        val voicePingThread = HandlerThread("VoicePingThread", Thread.MAX_PRIORITY)
        voicePingThread.start()
        player = Player(context, audioParam, DEFAULT_SERVER_URL, voicePingThread.looper)
        connection = OkConnection(context, player, voicePingThread.looper)
        recorder = Recorder(context, connection, audioParam, voicePingThread.looper)
        connection.setOutgoingAudioListener(recorder)
    }

    /**
     * Connect to server. This method can be assumed as sign in to server. After the user connected
     * to server, the user can then receive PTT from any other user using private channel.
     *
     * @param userId   User ID or Username
     * @param callback Callback
     */
    fun connect(userId: String, company: String, callback: ConnectCallback) {
        connect(DEFAULT_SERVER_URL, userId, company, callback)
    }

    /**
     * Connect to server. This method can be assumed as sign in to server. After the user connected
     * to server, the user can then receive PTT from any other user using private channel.
     *
     * @param serverUrl Server URL
     * @param userId    User ID or Username
     * @param callback  Callback
     */
    fun connect(serverUrl: String, userId: String, company: String, callback: ConnectCallback) {
        val realServerUrl = when {
            serverUrl.isBlank() -> DEFAULT_SERVER_URL
            serverUrl.endsWith("/") -> serverUrl.substring(0, serverUrl.length - 1)
            else -> serverUrl
        }
        this.userId = userId
        this.company = company
        val deviceId =
            "${Build.MANUFACTURER}_${Build.MODEL}_${Build.FINGERPRINT}_${Build.BOOTLOADER}_${Build.DISPLAY}_${Build.HOST}"
        player.setServerUrl(realServerUrl)
        connection.connect(realServerUrl, getFullUserId(), deviceId, callback)
        recorder.setUserId(getFullUserId())
    }

    /**
     * Disconnect from server. This method can be assumed as sign out from server. After
     * disconnected from server, the user will not receive any incoming message.
     *
     * @param callback Callback
     */
    fun disconnect(callback: DisconnectCallback) {
        connection.disconnect(callback)
    }

    /**
     * Get current connection state.
     *
     * @return Connection state
     */
    fun getConnectionState(): ConnectionState {
        return connection.connectionState
    }

    /**
     * Set ConnectionStateListener to VoicePing
     *
     * @param listener ConnectionStateListener
     */
    fun setConnectionStateListener(listener: ConnectionStateListener?) {
        connection.setConnectionStateListener(listener)
    }

    /**
     * Set IncomingTalkListener to VoicePing to do some advanced techniques to the incoming audio
     * data.
     *
     * @param listener IncomingTalkListener
     */
    fun setIncomingTalkListener(listener: IncomingTalkListener?) {
        player.setIncomingTalkListener(listener)
    }

    /**
     * Start PTT Talk using VoicePing's PTT functionality.
     *
     * @param receiverId  Receiver ID for private channel, or Group ID for group channel
     * @param channelType ChannelType.PRIVATE or ChannelType.GROUP
     * @param callback    OutgoingTalkCallback
     */
    fun startTalking(receiverId: String, channelType: Int, callback: OutgoingTalkCallback?) {
        recorder.startTalking(getFullId(receiverId), channelType, callback, null, null)
    }

    /**
     * Start PTT Talk using VoicePing's PTT functionality. Save recorded audio to the destination
     * path at the end of the talk. The saved file can be played using VoicePingPlayer.
     *
     * @param receiverId      Receiver ID for private channel, or Group ID for group channel
     * @param channelType     ChannelType.PRIVATE or ChannelType.GROUP
     * @param callback        OutgoingTalkCallback
     * @param destinationPath Destination path
     */
    fun startTalking(
        receiverId: String,
        channelType: Int,
        callback: OutgoingTalkCallback?,
        destinationPath: String?
    ) {
        recorder.startTalking(getFullId(receiverId), channelType, callback, destinationPath, null)
    }

    /**
     * Start PTT Talk using VoicePing's PTT functionality. Save recorded audio to the destination
     * path at the end of the talk. The saved file can be played using VoicePingPlayer. Add custom
     * audio recorder using CustomAudioRecorder.
     *
     * @param receiverId      Receiver ID for private channel, or Group ID for group channel
     * @param channelType     ChannelType.PRIVATE or ChannelType.GROUP
     * @param callback        OutgoingTalkCallback
     * @param destinationPath Destination path
     * @param recorder        CustomAudioRecorder
     */
    fun startTalking(
        receiverId: String,
        channelType: Int,
        callback: OutgoingTalkCallback?,
        destinationPath: String?,
        recorder: CustomAudioRecorder?
    ) {
        this.recorder.startTalking(
            getFullId(receiverId),
            channelType,
            callback,
            destinationPath,
            recorder
        )
    }

    /**
     * Stop PTT Talk.
     */
    fun stopTalking() {
        recorder.stopTalking()
    }

    /**
     * Join a group channel.
     *
     * @param groupId Group ID
     */
    fun joinGroup(groupId: String) {
        connection.send(MessageHelper.createSubscribeMessage(getFullUserId(), getFullId(groupId)))
    }

    /**
     * Leave a group channel.
     *
     * @param groupId Group ID
     */
    fun leaveGroup(groupId: String) {
        connection.send(MessageHelper.createUnsubscribeMessage(getFullUserId(), getFullId(groupId)))
    }

    /**
     * Get AudioParam that is used in this VoicePing instance.
     *
     * @return AudioParam
     */
    fun getAudioParam(): AudioParam {
        return audioParam
    }

    fun setAudioParam(audioParam: AudioParam) {
        this.audioParam = audioParam
        player.setAudioParam(audioParam)
        recorder.setAudioParam(audioParam)
    }

    /**
     * Mute from specific channel.
     *
     * @param targetId    Sender ID (Private) or Group ID (Group)
     * @param channelType ChannelType.PRIVATE or ChannelType.GROUP
     */
    fun mute(targetId: String, channelType: Int) {
        player.mute(getFullId(targetId), channelType)
    }

    /**
     * Mute all channels.
     */
    fun muteAll() {
        player.muteAll()
    }

    /**
     * Unmute to specific channel.
     *
     * @param targetId    Sender ID (Private) or Group ID (Group)
     * @param channelType ChannelType.PRIVATE or ChannelType.GROUP
     */
    fun unmute(targetId: String, channelType: Int) {
        player.unmute(getFullId(targetId), channelType)
    }

    /**
     * Unmute all channels.
     */
    fun unmuteAll() {
        player.unmuteAll()
    }

    private fun getFullUserId(): String {
        return "${company}_$userId"
    }

    private fun getFullId(id: String): String {
        return "${company}_$id"
    }
}